src/mscorlib/src/System/Threading/Tasks/Task.cs

```diff
@@ -5994,7 +5994,10 @@ public static Task WhenAll(IEnumerable<Task> tasks)
             foreach (Task task in tasks)
             {
                 if (task == null) ThrowHelper.ThrowArgumentException(ExceptionResource.Task_MultiTaskContinuation_NullTask, ExceptionArgument.tasks);
-                taskList.Add(task);
+                if (!task.IsRanToCompletion) // no need to wait on the task if it's already successfully completed
```

Comment:
> It looks like previously it would have transferred wait notifications from completed tasks to the when-all task here

Ugh, you're right, I forgot about that debugging aspect here.  The gain of this change is small enough that I'm not sure it's worth losing that.  We could probably regain it by checking whether the wait notification is set after checking if it's completed, but that's a larger thing to explore.  I'm going to close this.  Thanks for spotting that.